File: numsSameConsecDiff.py
#Approach 1- DFS [using call-stack]
def numsSameConsecDiff(n,k): # TC O(2^n) // SC O(2^n)
	if n == 1:
	    return [i for i in range(10)]

	ans = []
	def DFS(n, num):
	    # base case
	    if n == 0:
	        return ans.append(num)

	    tail_digit = num % 10
	    # using set() to avoid duplicates when K == 0
	    next_digits = set([tail_digit + k, tail_digit - k])

	    for next_digit in next_digits:
	        if 0 <= next_digit < 10:
	            new_num = num * 10 + next_digit
	            DFS(n-1, new_num)

	for num in range(1, 10):
	    DFS(n-1, num)

	return list(ans)

#Approach 2- BFS [using queue]
def numsSameConsecDiff1(n,k): # TC O(2^n) // SC O(2^n)
	if n == 1:
		return [i for i in range(10)]

	# initialize the queue with candidates for the first level
	queue = [digit for digit in range(1, 10)]

	for level in range(n-1):
	    next_queue = []
	    for num in queue:
	        tail_digit = num % 10
	        # using set() to avoid duplicates when K == 0
	        next_digits = set([tail_digit + k, tail_digit - k])

	        for next_digit in next_digits:
	            if 0 <= next_digit < 10:
	                new_num = num * 10 + next_digit
	                next_queue.append(new_num)
	    # start the next level
	    queue = next_queue

	return queue

print(numsSameConsecDiff(3,7))
print(numsSameConsecDiff1(3,7))

File: numsSameConsecDiff.java
import java.util.*;
// Approach 1- DFS (Recursive) [call-stack]
class Solution { // TC O(2^n) SC O(2^n)
    public List<Integer> numsSameConsecDiff(int N, int K) {
        if (N == 1)
            return new ArrayList<>(Arrays.asList(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)) ;

        List<Integer> results = new ArrayList<Integer>();
        for (int num = 1; num < 10; ++num)
            this.DFS(N - 1, num, K, results);

        // convert the ArrayList to int[]
        //return results.stream().mapToInt(i->i).toArray();
        return results;
    }

    protected void DFS(int N, int num, int K, List<Integer> results) {
        if (N == 0) {
            results.add(num);
            return;
        }
        List<Integer> nextDigits = new ArrayList<>();

        Integer tailDigit = num % 10;
        nextDigits.add(tailDigit + K);
        if (K != 0)
            nextDigits.add(tailDigit - K);
        for (Integer nextDigit : nextDigits) {
            if (0 <= nextDigit && nextDigit < 10) {
                Integer newNum = num * 10 + nextDigit;
                this.DFS(N - 1, newNum, K, results);
            }
        }
    }
}

// Approach 2- BFS [using queues] 
class Solution1{ // TC O(2^n) SC O(2^n)
    public List<Integer> numsSameConsecDiff(int N, int K) {

        if (N == 1)
            return new ArrayList<>(Arrays.asList(0,1,2,3,4,5,6,7,8,9)); 

        List<Integer> queue = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9);
        for(int level = 1; level < N; ++ level) {
            ArrayList<Integer> nextQueue = new ArrayList<>();
            // iterate through each number within the level
            for (Integer num : queue) {
                Integer tailDigit = num % 10;

                ArrayList<Integer> nextDigits = new ArrayList<>();
                nextDigits.add(tailDigit + K);
                if (K != 0)
                    nextDigits.add(tailDigit - K);
                for (Integer nextDigit : nextDigits) {
                    if (0 <= nextDigit && nextDigit < 10) {
                        Integer newNum = num * 10 + nextDigit;
                        nextQueue.add(newNum);
                    }
                }
            }
            // prepare for the next level
            queue = nextQueue;
        }

        //return queue.stream().mapToInt(i->i).toArray();
        return queue;
    }
}


class numsSameConsecDiff{
    public static void main(String[] args) {
        int n=3;
        int k=7;
        Solution s= new Solution();
        Solution1 s1= new Solution1();
        System.out.println(s.numsSameConsecDiff(n,k));
        System.out.println(s1.numsSameConsecDiff(n,k));
    }

}
